semgrep-core/Matching/Generic_vs_generic.ml

```diff
@@ -1643,143 +1654,152 @@ and m_list__m_stmt_uncached ~flattened (xsa: A.stmt list) (xsb: A.stmt list) =
-          )))
-
-  | A.While(a0, a1, a2), B.While(b0, b1, b2) ->
-      m_tok a0 b0 >>= (fun () ->
-        m_expr a1 b1 >>= (fun () ->
-          m_stmt a2 b2
-        ))
-  (*s: [[Generic_vs_generic.m_stmt]] boilerplate cases *)
-  | A.DefStmt(a1), B.DefStmt(b1) ->
-      m_definition a1 b1
-  | A.DirectiveStmt(a1), B.DirectiveStmt(b1) ->
-      m_directive a1 b1
-  (*x: [[Generic_vs_generic.m_stmt]] boilerplate cases *)
-
-  | A.DoWhile(a0, a1, a2), B.DoWhile(b0, b1, b2) ->
-      m_tok a0 b0 >>= (fun () ->
-        m_stmt a1 b1 >>= (fun () ->
-          m_expr a2 b2
-        ))
-  | A.For(a0, a1, a2), B.For(b0, b1, b2) ->
-      m_tok a0 b0 >>= (fun () ->
-        m_for_header a1 b1 >>= (fun () ->
-          m_stmt a2 b2
-        ))
-  | A.Switch(at, a1, a2), B.Switch(bt, b1, b2) ->
-      m_tok at bt >>= (fun () ->
-        m_option (m_expr) a1 b1 >>= (fun () ->
-          m_case_clauses a2 b2
-        ))
-
-  | A.Continue(a0, a1, asc), B.Continue(b0, b1, bsc) ->
-      let* () = m_tok a0 b0 in
-      let* () = m_label_ident a1 b1 in
-      m_tok asc bsc
-  | A.Break(a0, a1, asc), B.Break(b0, b1, bsc) ->
-      let* () = m_tok a0 b0 in
-      let* () = m_label_ident a1 b1 in
-      m_tok asc bsc
-  | A.Label(a1, a2), B.Label(b1, b2) ->
-      m_label a1 b1 >>= (fun () ->
-        m_stmt a2 b2
-      )
-  | A.Goto(a0, a1), B.Goto(b0, b1) ->
-      m_tok a0 b0 >>= (fun () ->
-        m_label a1 b1
-      )
-  | A.Throw(a0, a1, asc), B.Throw(b0, b1, bsc) ->
-      let* () = m_tok a0 b0 in
-      let* () = m_expr a1 b1 in
-      m_tok asc bsc
-  | A.Try(a0, a1, a2, a3), B.Try(b0, b1, b2, b3) ->
-      let* () = m_tok a0 b0 in
-      let* () = m_stmt a1 b1 in
-      let* () = (m_list m_catch) a2 b2 in
-      (m_option m_finally) a3 b3
-  | A.Assert(a0, a1, a2, asc), B.Assert(b0, b1, b2, bsc) ->
-      let* () = m_tok a0 b0 in
-      let* () = m_expr a1 b1 in
-      let* () = (m_option m_expr) a2 b2 in
-      m_tok asc bsc
-
-  | A.OtherStmt(a1, a2), B.OtherStmt(b1, b2) ->
-      m_other_stmt_operator a1 b1 >>= (fun () ->
-        (m_list m_any) a2 b2
-      )
-  | A.OtherStmtWithStmt(a1, a2, a3), B.OtherStmtWithStmt(b1, b2, b3) ->
-      m_other_stmt_with_stmt_operator a1 b1 >>= (fun () ->
-        m_option m_expr a2 b2 >>= (fun () ->
-          m_stmt a3 b3
-        ))
-  | A.WithUsingResource(a1, a2, a3), B.WithUsingResource(b1, b2, b3) ->
-      m_tok a1 b1 >>= (fun () ->
-        m_stmt a2 b2 >>= (fun () ->
-          m_stmt a3 b3
-        ))
-
-  | A.ExprStmt _, _  | A.DefStmt _, _  | A.DirectiveStmt _, _
-  | A.Block _, _  | A.If _, _  | A.While _, _  | A.DoWhile _, _  | A.For _, _
-  | A.Switch _, _  | A.Return _, _  | A.Continue _, _  | A.Break _, _
-  | A.Label _, _  | A.Goto _, _  | A.Throw _, _  | A.Try _, _
-  | A.Assert _, _
-  | A.OtherStmt _, _ | A.OtherStmtWithStmt _, _ | A.WithUsingResource _, _
-    -> fail ()
+  let pattern_strs = Bloom_annotation.list_of_pattern_strings (S a) in
```

Comment:
No! I don't think it's a good idea to do this each time in m_stmt.
Either we also pre-compute also on the pattern the bloom filter, and set it in the s_bf field of the pattern,
or we leave the logic outside Generic_vs_generic. Generic_vs_generic is already quite complicated
so I would really prefer to have the optimizations done outside, e.g., in Semgrep_generic.ml as it was before.

Reply:
This was an experiment---I was planning on asking you for advice on Monday, but I'm not getting any speedup on stmt lists because it doesn't check the pattern against a bloom filter first. See my reply to Martin in eng.